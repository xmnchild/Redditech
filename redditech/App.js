import * as React from "react";
import * as WebBrowser from "expo-web-browser";

//* ------- Components ------- *//
import Route from "./routes";

//* ------- Recoil ------- *//
import { RecoilRoot } from "recoil";

WebBrowser.maybeCompleteAuthSession();

export default function App() {
  // removeToken("secure_token");
  return (
    <RecoilRoot>
      <Route />
    </RecoilRoot>
  );
}
