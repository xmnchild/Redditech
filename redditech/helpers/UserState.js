import { atom } from "recoil";
///La liste des devices Ã  utiliser au travers de la map
const user = atom({ key: "userState", default: undefined });
const tokenAtom = atom({ key: "token", default: undefined });

const settingsAtom = atom({ key: "settings", default: {} });

export { user, tokenAtom, settingsAtom };
