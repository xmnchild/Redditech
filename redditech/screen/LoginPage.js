import { useEffect } from "react";
import { StyleSheet, Text, View, Image, Pressable, Linking } from "react-native";
import { useNavigation } from "@react-navigation/native";
import { useAuthRequest } from "expo-auth-session";
import * as WebBrowser from "expo-web-browser";
import { encode as btoa } from "base-64";
import axios from "axios";
//* ------- RECOIL ------- *//
import { useRecoilState } from "recoil";
//* ------- STORE ------- *//
import { setToken } from "../store/token";
//* ------- ATOMS ------- *//
import { tokenAtom } from "../helpers/UserState";
WebBrowser.maybeCompleteAuthSession();
const CLIENT_ID = "8H3_j3pSQaxgfHJkG6hwlA";
// Endpoint
const discovery = {
  authorizationEndpoint: "https://www.reddit.com/api/v1/authorize",
  tokenEndpoint: "https://www.reddit.com/api/v1/access_token",
};
export default function LoginPage() {
  const navigation = useNavigation();
  const [request, response, promptAsync] = useAuthRequest(
    {
      clientId: CLIENT_ID,
      scopes: ["*"],
      redirectUri: "exp://localhost:19000/",
    },
    discovery
  );
  const [, setDoneToken] = useRecoilState(tokenAtom);
  //function to get token
  const getToken = async () => {
    if (response?.type === "success") {
      const { code } = response.params;
      const base64 = btoa(`${CLIENT_ID}:`);
      try {
        const data = await axios.post(
          "https://www.reddit.com/api/v1/access_token",
          `grant_type=authorization_code&code=${code}&redirect_uri=exp://localhost:19000/`,
          {
            headers: {
              Authorization: `Basic ${base64}`,
              "content-type": "application/x-www-form-urlencoded",
            },
          }
        );
        const token = data["data"]["access_token"];
        setDoneToken(token);
        setToken(token);
        navigation.navigate("Feed");
      } catch (error) {
        console.log(error);
      }
    }
  };
  useEffect(() => {
    getToken();
  }, [response]);
  return (
    <View style={{ flex: 1, marginTop: 100, alignItems: "center" }}>
      <Image
        source={require("../assets/redditech-logo.png")}
        style={styles.logo}
      />
      {/* <Text style={styles.title}>Welcome to Hcetidder!</Text> */}
      <View style={{ flexGrow: 1 }} />
      <Pressable
        disabled={!request}
        title="Login"
        onPress={() => {
          promptAsync();
        }}
        style={styles.button}>
        <Text style={styles.buttonText}>Log in with your Reddit account</Text>
      </Pressable>

      <View style={{ marginTop: 10, marginBottom: 100, flexDirection: "row" }}>
        <Text style={{ fontSize: 16 }}>New to Reddit? </Text>
        <Pressable style={{ padding: 0 }} onPress={onPressSignup}>
          <Text style={styles.registerText}>Sign up</Text>
        </Pressable>
      </View>
    </View>
  );
}
async function onPressSignup(navigation) {
  Linking.addEventListener("url", handleUrlChange);
  await Linking.openURL("https://www.reddit.com/register/");
  function handleUrlChange(event) {
    const { url } = event;
    // check if the URL matches the registration complete URL
    if (url && url.startsWith("https://www.reddit.com/onboarding")) {
      Linking.removeEventListener("url", handleUrlChange);
      navigation.goBack();
    }
  }
}
const styles = StyleSheet.create({
  logo: {
    width: 250,
    height: 250,
    marginVertical: 50,
  },
  container: {
    padding: 10,
    flex: 1,
  },
  button: {
    paddingHorizontal: 40,
    paddingVertical: 10,
    backgroundColor: "#ff5700",
    borderRadius: 40,
  },
  buttonText: {
    color: "white",
    fontSize: 20,
  },
  title: {
    marginBottom: 30,
    fontSize: 24,
    fontWeight: 'bold'
  },
  registerText: {
    fontSize: 16,
    color: "navy",
  },
});
