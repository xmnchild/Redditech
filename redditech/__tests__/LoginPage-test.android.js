import * as React from "react";
import LoginPage from "../screen/LoginPage"
import renderer from "react-test-renderer";
import { RecoilRoot } from "recoil";
import { NavigationContainer } from "@react-navigation/native";
import { tokenAtom, settingsAtom } from "../helpers/UserState";


/*
//jest.useFakeTimers()
it(`renders correctly`, () => {
    const tree = renderer.create(<RecoilRoot><NavigationContainer><LoginPage/></NavigationContainer></RecoilRoot>);
    expect(tree).toMatchSnapshot();
});
*/
import App from '../App';

describe('<App />', () => {
    it('has 1 child', () => {
        const tree = renderer.create(<App />).toJSON();
        expect(tree.children.length).toBe(1);
    });
});
