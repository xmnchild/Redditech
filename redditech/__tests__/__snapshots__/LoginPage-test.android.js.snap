// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "marginTop": 100,
    }
  }
>
  <Image
    source={1}
    style={
      Object {
        "height": 250,
        "marginVertical": 50,
        "width": 250,
      }
    }
  />
  <View
    style={
      Object {
        "flexGrow": 1,
      }
    }
  />
  <View
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#ff5700",
        "borderRadius": 40,
        "paddingHorizontal": 40,
        "paddingVertical": 10,
      }
    }
    title="Login"
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 20,
        }
      }
    >
      Log in with your Reddit account
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 100,
        "marginTop": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 16,
        }
      }
    >
      New to Reddit? 
    </Text>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "padding": 0,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "navy",
            "fontSize": 16,
          }
        }
      >
        Sign up
      </Text>
    </View>
  </View>
</View>
`;
